fgpyo/sam/__init__.py

```diff
@@ -934,3 +949,111 @@ class SamOrder(enum.Enum):
     Coordinate = "coordinate"  #: coordinate sorted
     QueryName = "queryname"  #: queryname sorted
     Unknown = "unknown"  # Unknown SAM / BAM / CRAM sort order
+
+
+def extract_umis_from_read_name(
+    read_name: str,
+    read_name_delimiter: str = ILLUMINA_READ_NAME_DELIMITER,
+    umi_delimiter: str = ILLUMINA_UMI_DELIMITER,
+    strict: bool = False,
+) -> Optional[str]:
+    """Extract UMI(s) from a read name.
+
+    The UMI is expected to be the final component of the read name, delimited by the
+    `read_name_delimiter`. Multiple UMIs may be present, delimited by the `umi_delimiter`. This
+    delimiter will be replaced by the SAM-standard `-`.
+
+    Args:
+        read_name: The read name to extract the UMI from.
+        read_name_delimiter: The delimiter separating the components of the read name.
+        umi_delimiter: The delimiter separating multiple UMIs.
+        strict: If `strict` is true, the read name must contain either 7 or 8 colon-separated
+          segments. The UMI is assumed to be the last one in the case of 8 segments and `None`
+          in the case of 7 segments. `strict` requires the UMI to be valid and consistent with
+          Illumina's allowed UMI characters. If `strict` is false, the last segment is returned
+          so long as it appears to be a valid UMI.
+
+    Returns:
+        The UMI extracted from the read name, or None if no UMI was found. Multiple UMIs are
+        returned in a single string, separated by a hyphen (`-`).
+
+    Raises:
+        ValueError: If the read name does not end with a valid UMI.
+    """
+    if strict:
+        colons = read_name.count(":")
+        if colons == 6:  # number of fields is 7
+            return None
+        elif colons != 7:
+            raise ValueError(
+                f"Trying to extract UMIs from read with {colons + 1} parts "
+                f"(7 or 8 expected): {read_name}"
+            )
+    raw_umi = read_name.split(read_name_delimiter)[-1]
+    # Check each UMI individually
+    umis = raw_umi.split(umi_delimiter)
+    # Strip the "r" from rev-comped UMIs
+    # (NB: for consistency with UMI_tools, the UMI is not revcomped)
+    umis = [umi.lstrip("r") for umi in umis]
+
+    invalid_umis = [umi for umi in umis if not _is_valid_umi(umi)]
+    if len(invalid_umis) > 0:
+        if strict:
+            raise ValueError(
+                f"Invalid UMIs found in read name: {read_name}",
+                f"  (Invalid UMIs: {', '.join(invalid_umis)})",
+            )
+        else:
+            return None
+
+    else:
+        return SAM_UMI_DELIMITER.join(umis)
```

Comment:
I think we can re-order this to remove the nesting.  Strictly speaking, we could change the `elif` to an `if`, and then remove the `else`, but I think this is clearer:

```suggestion
    if len(invalid_umis) == 0:
        return SAM_UMI_DELIMITER.join(umis)
    elif strict:
        raise ValueError(
            f"Invalid UMIs found in read name: {read_name}",
            f"  (Invalid UMIs: {', '.join(invalid_umis)})",
        )
    else:
        return None
```